### PR TITLE
add caseDataSource and StartUpRunner

### LINKED ISSUE #18: CaseServerDataSource
* **Do you want to request a *feature* or report a *bug*?**
Feature

* **What is the current behavior?**
Connection between REST API and IIDM data source is not easy. 

* **What is the expected behavior?**
We have to implement a `CaseServerDataSource':

```java
class CaseServerDataSource implement ReadOnlyDataSource {

    public CaseServerDataSource(RestTemplate restTemplate) {
    }
...
}
```

That will connect to case server REST API to get data streams. Following new methods have to be implemented.

```http
/cases/:caseName?fileName=
/cases/:caseName?suffix=&ext=
/cases/:caseName/exists?fileName=
/cases/:caseName/exists?suffix=&ext=
```

We also to have to pay attention in the implementation so that data is streams from end to end (from file on case server side, to importer that uses the data source).
Also byte stream in the REST API could be compressed (could be done later on).

* **What is the motivation / use case for changing the behavior?**

* **Please tell us about your environment:**
  - PowSyBl Version: ...
  - OS Version: ...

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, spectrum, etc)

(if a question doesn't apply, you can delete it)

### PR DESCRIPTION
**Pull request by AbdelHedhili**
Signed-off-by: AbdelHedhili <abdelsalem.hedhili@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
Fix #18 and #8

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -53,9 +53,10 @@
         <maven.surefire.version>2.22.0</maven.surefire.version>
 
         <google.guava.version>20.0</google.guava.version>
+        <google.gson.version>2.8.6</google.gson.version>
```
Comment (geofjamg) [inline pom.xml]: Why are you using an additional json framework? Jackson is already used internally by Spring.

File: src/main/java/com/powsybl/caseserver/datasource/util/CaseDataSourceService.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ */
+@Service
+@ComponentScan(basePackageClasses = CaseService.class)
+public class CaseDataSourceService {
+
+    @Autowired
+    private CaseService caseService;
+
+    String getBaseName(String caseName) {
+        caseService.checkStorageInitialization();
```
Comment (geofjamg) [inline src/main/java/com/powsybl/caseserver/datasource/util/CaseDataSourceService.java]: As this 2 lines of code are used everywhere you can have add a getDataSource or something like this that share this 2 lines.

File: src/main/java/com/powsybl/caseserver/datasource/util/CaseDataSourceService.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        return dataSource.getBaseName();
+    }
+
+    Boolean datasourceExists(String caseName, String suffix, String ext) {
+        caseService.checkStorageInitialization();
+        DataSource dataSource = initDatasource(caseName);
+        try {
+            return dataSource.exists(suffix, ext);
+        } catch (IOException e) {
+            throw new CaseException(e.getMessage());
```
Comment (geofjamg) [inline src/main/java/com/powsybl/caseserver/datasource/util/CaseDataSourceService.java]: A general remark, for IOException there is already a UncheckedException to wrap it with a runtime exception
=> throw new UncheckIOException(e);

File: src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                fail();
+            }
+        }
+        caseService.setFileSystem(fileSystem);
+        //insert a cgmes in the FS
+        try (InputStream cgmesURL = getClass().getResourceAsStream("/" + cgmesName)) {
+            Path cgmes = path.resolve(cgmesName);
+            Files.copy(cgmesURL, cgmes, StandardCopyOption.REPLACE_EXISTING);
+        }
+        dataSource = Importers.createDataSource(Paths.get(getClass().getResource("/" + cgmesName).toURI()));
```
Comment (geofjamg) [inline src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java]: Use ResourceDataSource instead. It has been designed to easy class resource loading as data source.

File: src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    @Test
+    public void testListName() throws Exception {
+        MvcResult mvcResult = mvc.perform(get("/v1/cases/{caseName}/datasource/list", cgmesName)
+                .param("regex", ".*"))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        Gson gson = new GsonBuilder()
```
Comment (geofjamg) [inline src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java]: I do not see the reason to use this. There is already a way with Spring MVC mock to assert json result.

File: src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        dataSource = Importers.createDataSource(Paths.get(getClass().getResource("/" + cgmesName).toURI()));
+    }
+
+    @Test
+    public void testBaseName() throws Exception {
+        MvcResult mvcResult = mvc.perform(get("/v1/cases/{caseName}/datasource/baseName", cgmesName))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        assertEquals(dataSource.getBaseName(), mvcResult.getResponse().getContentAsString());
```
Comment (geofjamg) [inline src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java]: Why using an asserEquals? You could just do it with a andExpect(...) ?

File: src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        MvcResult mvcResult = mvc.perform(get("/v1/cases/{caseName}/datasource", cgmesName)
+                .param("fileName", fileName))
+                .andExpect(request().asyncStarted())
+                .andReturn();
+
+        mvcResult = mvc.perform(asyncDispatch(mvcResult))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        InputStreamReader isReader = new InputStreamReader(dataSource.newInputStream(fileName));
```
Comment (geofjamg) [inline src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java]: Inputstreams readers have to be closed.

File: src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java (hunk 6 above)
Comment (AbdelHedhili) [reply src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java]: I'd rather let the mvc perform to check everything about the status and check the result after with the classic assert. It's just for readability, I don't like to put everything in a block. We don't lose or earn anything either way anyway.

Comment (geofjamg) [reply src/test/java/com/powsybl/caseserver/datasource/util/CaseDataSourceControllerTest.java]: ok why not

File: powsybl-case-datasource-client/pom.xml [hunk 8]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (geofjamg) [inline powsybl-case-datasource-client/pom.xml]: Add license header

File: powsybl-case-datasource-client/src/main/java/com/powsybl/cases/datasource/CaseDataSourceClientException.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+/** Copyright (c) 2020, RTE (http://www.rte-france.com)
```
Comment (geofjamg) [inline powsybl-case-datasource-client/src/main/java/com/powsybl/cases/datasource/CaseDataSourceClientException.java]: This is a missing line break

File: powsybl-case-distribution/pom.xml [hunk 10]
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (geofjamg) [inline powsybl-case-distribution/pom.xml]: Add license header

File: powsybl-case-server/pom.xml [hunk 11]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (geofjamg) [inline powsybl-case-server/pom.xml]: Add license header

**Pull request merged**